lib/pharos/up_command.rb

```diff
@@ -152,42 +141,39 @@ def worker_hosts(config)
       config.hosts.select { |h| h.role == 'worker' }
     end
 
-    # @param master [Pharos::Configuration::Node]
-    # @param features [Pharos::Config]
-    def handle_masters(master, config)
-      log_host_header(master)
-      Phases::ConfigureHost.new(master).call
-      Phases::ConfigureMaster.new(master, config).call
-      Phases::ConfigureClient.new(master).call
-      Phases::ConfigureDNS.new(master, config).call
-      Phases::ConfigureNetwork.new(master, config.network).call
-      Phases::ConfigureMetrics.new(master).call
-      Phases::LabelNode.new(master, master).call
-      Phases::StoreClusterYAML.new(master, config_yaml.read(ENV.to_h)).call
+    def handle_phases(config)
+      all_hosts = config.hosts
+      master_hosts = master_hosts(config)
+      master_host = master_hosts[0]
+      worker_hosts = worker_hosts(config)
+
+      handle_phase(Phases::ValidateHost, all_hosts, ssh: true, parallel: true)
+      handle_phase(Phases::ConfigureHost, all_hosts, ssh: true, parallel: true)
+
+      handle_phase(Phases::ConfigureMaster, master_hosts, ssh: true)
+      handle_phase(Phases::ConfigureClient, master_hosts, ssh: true)
+
+      # master is now configured and can be used
+      handle_phase(Phases::ConfigureDNS, [master_host], master: master_host)
+      handle_phase(Phases::ConfigureNetwork, [master_host], master: master_host)
+      handle_phase(Phases::ConfigureMetrics, [master_host], master: master_host)
```

Comment:
These three phases don't actually use the `@host` at all... they only use the `@config` and `Pharos::Kube.client(@master.address)`. The `[master_host]` is only used for the log output.